### PR TITLE
BUG: Fix bug in loc setitem changing the dtype when condition is False

### LINKED ISSUE #37550: BUG: Loc changes dtype when condition is completly False
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
df = pd.DataFrame({
    "a": ["a"],
    "b": [1],
    "c": [1]
})
df.loc[[False], ["b"]] = 10 - df["c"]
print(df)

```

Thus converts ``b`` to float and returns 

```
   a    b  c
0  a  1.0  1
```

#### Problem description

[this should explain **why** the current behaviour is a problem and why the expected output is a better solution]

#### Expected Output

Would expect that this returns an integer.

```
   a    b  c
0  a  1  1
```

Interestingly, this is returned if we drop the column ``a``.
```
df = pd.DataFrame({
    "b": [1],
    "c": [1]
})
df.loc[[False], ["b"]] = 10 - df["c"]
print(df)
```

or if we are not using df["c"] and an integer instead.
```
df = pd.DataFrame({
    "a": ["a"],
    "b": [1],
    "c": [1]
})
df.loc[[False], ["b"]] = 10 - 1
print(df)
```

Both return ``1`` instead of ``1.0`` for column ``b``

Edit: The condition [False] is just a simplification. This also happens if we use a condition, which evaluates sometimes to completly False based on the input data.

#### Output of ``pd.show_versions()``

<details>

master

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #37550
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This fixes the bug, but I am not sure, if the solution is good enough....

### CODE REVIEW

File: pandas/core/internals/blocks.py
Comment (jreback) [inline pandas/core/internals/blocks.py:926]: can you put this as another elif case

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py
Comment (phofl) [reply pandas/core/internals/blocks.py]: Done

File: pandas/core/internals/blocks.py
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:923]: any way to avoid the np.array call?  i think that makes a copy if we have eg a list

File: pandas/tests/frame/indexing/test_setitem.py
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_setitem.py:301]: can you put this below in the class TestDataFrameSetItemBooleanMask?

which part(s) of the line `df.loc[[False], ["b"]] = 10 - df["c"]` are relevant to the bugfix here? e.g. does it matter that the right hand side is a Series?  Would this be testing the right thing if it were `series.loc[[False]]`?

Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py:301]: ``df.loc[[False], ["b"]] = 10`` did not trigger the error. We need a Series on the rhs.

File: pandas/core/internals/blocks.py
Comment (phofl) [reply pandas/core/internals/blocks.py:923]: I don't think so, ``is_empty_indexer`` requires an ndarray, while may be a list.

File: pandas/tests/frame/indexing/test_setitem.py
Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py:301]: Moved the test

File: pandas/tests/frame/indexing/test_setitem.py
Comment (jreback) [inline pandas/tests/frame/indexing/test_setitem.py:347]: can you *also* test the rhs as just a scalar integer (e.g. 10 - 1), apparently this does the same behavior.

Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py:347]: Done

Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py:347]: Runs a different code path, that is the reason we did not have problems with a scalar :)

File: doc/source/whatsnew/v1.2.0.rst
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.2.0.rst:472]: changed -> changing

these should usually be participles

File: pandas/core/internals/blocks.py
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:957]: the test is for `loc.__setitem__`.  is that (or i guess `iloc.__setitem__`) the only way we can get here?  it _may_ make sense to do this in setitem_with_indexer

File: pandas/tests/frame/indexing/test_setitem.py
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_setitem.py:343]: can you comment and/or edit the test name to make clear

- is iloc affected?
- does it matter that you have `["b"]` instead of, say, `"b"` or ":"?

comment that the bug is for the value being set being a Series obj

Comment (jbrockmendel) [reply pandas/tests/frame/indexing/test_setitem.py:343]: is Series affected?

File: doc/source/whatsnew/v1.2.0.rst
Comment (phofl) [reply doc/source/whatsnew/v1.2.0.rst:472]: Thx, will follow this in the future

File: pandas/tests/frame/indexing/test_setitem.py
Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py:343]: Series is not affected. Goes only wrong with ``["b"]`` 

``Iloc`` is actually also affected, but has to be fixed differently probably. In case of ``loc`` we get an empty indexer. In case of ``iloc`` we get an indexer like ``([False],)``, which is not empty per ``is_empty_indexer``. We could adjust ``is_empty_indexer`` or we have to go a different way here

File: pandas/core/internals/blocks.py
Comment (phofl) [reply pandas/core/internals/blocks.py:957]: Yeah, that would be possible I think.
Sport would be in https://github.com/pandas-dev/pandas/blob/524fc9c6790ddc4b151b6fb910a27073961c5fba/pandas/core/indexing.py#L1567

Ran the tests in indexing, frame/indexing and series/indexing locally, this seems to work. Let me know, if we should move this

File: pandas/tests/frame/indexing/test_setitem.py
Comment (jreback) [reply pandas/tests/frame/indexing/test_setitem.py:348]: sure that's fine, but just locking it down with a test

File: pandas/core/internals/blocks.py
Comment (jreback) [reply pandas/core/internals/blocks.py:931]: you can use np.asarray, also can we check len of the indexer here first to short-cut?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py
Comment (phofl) [reply pandas/core/internals/blocks.py:931]: Thx used asarray. Indexer is a tuple containing an empty list, len does not work here unfortunately.

File: pandas/tests/frame/indexing/test_setitem.py
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_setitem.py:351]: you can simplify to 9

File: pandas/tests/frame/indexing/test_setitem.py
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_setitem.py:347]: can you make clear either in the test name and/or comment that/whether the type of indexer for `["b"]` is relevant

File: pandas/core/internals/blocks.py
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:931]: im not sure this is the right place to catch this, since i think we are specifically interested in all-False boolean indexers coming through loc.  does this get here via setitem_with_indexer, and if so, can you point out specifically which branch(es)?

File: pandas/tests/frame/indexing/test_setitem.py
Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py]: Thx

File: pandas/tests/frame/indexing/test_setitem.py
Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py]: Added comment and parametrized test

File: pandas/core/internals/blocks.py
Comment (phofl) [reply pandas/core/internals/blocks.py:931]: Through ``_setitem_with_indexer_split_path``  -> ``_setitem_with_indexer_2d_value`` -> ``_setitem_single_column``

as mentioned in https://github.com/pandas-dev/pandas/pull/37672#discussion_r523415484 we could do that in ``setitem_with_indexer`` if this is better

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:931]: In _setitem_with_indexer_split_path L 1686-1688 we have a no-op block that is for all-false boolean masks.  could move it up to before the 2D check.  would that do the trick?

Comment (phofl) [reply pandas/core/internals/blocks.py]: One possible solution is moving up the ``lplane_indexer==0`` condition and removing the value part. Probably the shortest and does not seem to break tests. Don't know if this kills behavior not tested, but if indexer has len=0, we probably do not want to set values anyway

File: pandas/tests/frame/indexing/test_setitem.py
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_setitem.py:354]: should `[False]` be `func(value)` as above?  if not, this might belong as a separate test

File: pandas/tests/frame/indexing/test_setitem.py
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_setitem.py:343]: can you call this "box"

File: pandas/tests/frame/indexing/test_setitem.py
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_setitem.py:344]: it isnt clear to me that this is correct with `[]`.  why is it included?

File: pandas/core/indexing.py
Comment (jbrockmendel) [inline pandas/core/indexing.py:1686]: i think we need to retain the check for len(value) since we only want to catch boolean masking (though clearly we dont have tests with a mismatched len(value))

probably also need to check that pi is a listlike with matching length too.

Comment (jbrockmendel) [reply pandas/core/indexing.py:1686]: indexers.check_setitem_lengths may be useful here

File: pandas/tests/frame/indexing/test_setitem.py
Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py:344]: Shouldn't an empty indexer do nothing to the underlying object?

File: pandas/tests/frame/indexing/test_setitem.py
Comment (phofl) [reply pandas/tests/frame/indexing/test_setitem.py:343]: Done

File: pandas/core/indexing.py
Comment (phofl) [reply pandas/core/indexing.py:1686]: What would we want to do with empty indexers then?

Comment (phofl) [reply pandas/core/indexing.py:1678]: The reason I though this would be ok is, that empty indexers look here the same as all False boolean indexers. Value in is here empty too, so we have that mismatch

Comment (jbrockmendel) [reply pandas/core/indexing.py:1678]: i think `ser.loc[[]] = [1, 2, 3]` should raise kind of like `ser.loc[[1]] = [1, 2, 3, 4]` would

Comment (phofl) [reply pandas/core/indexing.py:1678]: Problem here is:

In case of ``[]`` we have the following parameters:

```
lplane_indexer=0
indexer = (array([], dtype=int64), array([1])
value = array([])
```

In case of ``[False]``  we get exactyl the same parametrization. We cant really decide here whether we got a boolean indexer or not.

Additionally,

```
df = DataFrame({"a": ["a"], "b": [1], "c": [1]})
df.loc[list([]), ["b"]] = 10 - df["c"]
df.loc[list([]), ["b"]] = 11
```

does not raise on master either. Was not introduced here

Comment (jbrockmendel) [reply pandas/core/indexing.py:1678]: That is a tough one.  Let's get another opinion.  @jreback: what do you expect to happen with `ser.loc[[]] = [1, 2, 3]`?

Comment (phofl) [reply pandas/core/indexing.py:1694]: ping @jreback ?

**Comment by simonjayhawkins** [comment]
@phofl can you merge master to resolve conflicts

**Comment by jreback** [comment]
@phofl can you rebase and will look again

**Comment by phofl** [comment]
merged master

File: pandas/core/indexing.py
Comment (jreback) [reply pandas/core/indexing.py:1741]: coming back to this, yeah i think if we don't have equal len indexer we should always raise, so @jbrockmendel example should raise

**Comment by jreback** [comment]
@jbrockmendel ok here?

**Comment by jreback** [comment]
@phofl can you update

**Comment by jbrockmendel** [comment]
Looks like the `ser[[]] = [1, 2, 3]` thing hasnt been addressed

**Comment by jreback** [comment]
ok let's move off the milestone for rn, can always do after the rc

**Comment by jreback** [comment]
@phofl status of this (merge master)

**Comment by jreback** [comment]
@phofl worth fixing up?

**Comment by phofl** [comment]
Yeah definitely.

i hope that I can get back to this friday or friday next week

**Comment by phofl** [comment]
I moved the fix for the all False case. This fixes an inconsistency in _align_series and covers the all False indexer case. 

I could not figure the empty list thing out. The values look exactly the same as in the all False case. We probably will have to fix this someplace else, but not sure right now.

I woild propose open an issue about that and go from there.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm @jbrockmendel if any comments

**Comment by jbrockmendel** [comment]
will take a look this afternoon.  (got booster yesterday, moving slow today)

File: pandas/core/indexing.py
Comment (jbrockmendel) [inline pandas/core/indexing.py:2062]: evidently it works, but this seems like a weird place to handle this.  is my intuition wrong here?

**Comment by jbrockmendel** [comment]
Does this relate to #43632?

File: pandas/core/indexing.py
Comment (phofl) [reply pandas/core/indexing.py:2062]: This ensures, that an indexer like ``(np.array([]), np.array([1]))`` is handled the same as ``(np.array([]), 1)``, hence I handled it here.

**Comment by phofl** [comment]
I don't think so. The all False case is an empty not a boolean indexer when jumping into the setitem logic

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**
